### PR TITLE
Support for Apache HttpClient 5 plugin

### PR DESCRIPTION
**Pull request by heihaozi**
### Support for Apache HttpClient 5 plug
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [ ] Update the documentation to include this new feature.
- [x] Tests(including UT, IT, E2E) are added to verify the new feature.
- [ ] If it's UI related, attach the screenshots below.

---

- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes [apache/skywalking#7763](https://github.com/apache/skywalking/issues/7763).
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking-java/blob/main/CHANGES.md).

### CODE REVIEW

**Comment by wu-sheng** [comment]
Please fix CI, and you need to follow plugin dev doc to add integration tests for new plugins, which makes sure this works in both old and new(5.x) releases.

**Comment by heihaozi** [comment]
OK, Does the integration test create the corresponding project in `\test\plugin\scenarios`?

**Comment by wu-sheng** [comment]
> OK, Does the integration test create the corresponding project in `\test\plugin\scenarios`?

Yes, do you know where the document is?

**Comment by heihaozi** [comment]
I don’t know, where is the document?

**Comment by wu-sheng** [comment]
1. https://skywalking.apache.org/docs/skywalking-java/latest/en/setup/service-agent/java-agent/java-plugin-development-guide/
2. https://skywalking.apache.org/docs/skywalking-java/latest/en/setup/service-agent/java-agent/plugin-test/ linked from (1).

**Comment by heihaozi** [comment]
I am using a windows system and cannot execute the `bash ./test/plugin/run.sh -f ${scenario_name}` command. 
Is there an alternative?

**Comment by wu-sheng** [comment]
> I am using a windows system and cannot execute the `bash ./test/plugin/run.sh -f ${scenario_name}` command.
> Is there an alternative?

It only works on Linux or MacOS. Could you find one?

**Comment by heihaozi** [comment]
I will try.

**Comment by wu-sheng** [comment]
`/docs/en/setup/service-agent/java-agent/Plugin-list.md` should be updated in every time you add a new plugin.

**Comment by wu-sheng** [comment]
Also `Supported-list.md`

**Comment by wu-sheng** [comment]
Your new case has not been added into GitHub action control file, so, no verification will happen in the CI process.

**Comment by heihaozi** [comment]
Is the `segments` in `expectedData.yaml` in order?

**Comment by wu-sheng** [comment]
> Is the `segments` in `expectedData.yaml` in order?

For segments, NO. Because those are not expected, mock traffic many trace segments, we just check whether there is at least one matched.

File: test/plugin/scenarios/httpclient-5.x-scenario/config/expectedData.yaml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+              - {key: url, value: 'http://127.0.0.1:8080/httpclient-5.x/case/asyncGet'}
+              - {key: http.method, value: GET}
+            skipAnalysis: 'false'
+          - operationName: /httpclient-5.x/case/get
+            parentSpanId: -1
+            spanId: 0
+            spanLayer: Http
+            startTime: nq 0
+            endTime: nq 0
+            componentId: not null
```
Comment (wu-sheng) [inline test/plugin/scenarios/httpclient-5.x-scenario/config/expectedData.yaml:113]: Please use real componentId in the expectedData file.

File: apm-sniffer/apm-sdk-plugin/httpclient-5.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/httpclient/v5/HttpAsyncClientDoExecuteInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+            MethodInterceptResult result) throws Throwable {
+        AsyncResponseConsumer consumer = (AsyncResponseConsumer) allArguments[2];
+        HttpContext context = (HttpContext) allArguments[4];
+        FutureCallback callback = (FutureCallback) allArguments[5];
+        allArguments[2] = new AsyncResponseConsumerWrapper(consumer);
+        allArguments[5] = new FutureCallbackWrapper(callback);
+        if (ContextManager.isActive() && !ContextManager.activeSpan().isExit()) {
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/httpclient-5.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/httpclient/v5/HttpAsyncClientDoExecuteInterceptor.java:43]: Why do we need to check `isExit`?

File: apm-sniffer/apm-sdk-plugin/httpclient-5.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/httpclient/v5/HttpClientDoExecuteInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            return uri;
+        }
+    }
+
+    private boolean isUrl(String uri) {
+        String lowerUrl = uri.toLowerCase();
+        return lowerUrl.startsWith("http") || lowerUrl.startsWith("https");
+    }
+
+    private String buildSpanValue(HttpHost httpHost, String uri) {
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/httpclient-5.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/httpclient/v5/HttpClientDoExecuteInterceptor.java:118]:
```suggestion
    private String buildURL(HttpHost httpHost, String uri) {
```

File: apm-sniffer/apm-sdk-plugin/httpclient-5.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/httpclient/v5/HttpAsyncClientDoExecuteInterceptor.java (hunk 2 above)
Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/httpclient-5.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/httpclient/v5/HttpAsyncClientDoExecuteInterceptor.java:43]: What is the case there is no `exit span` available? We usually use this kind of check when these are expected.

File: test/plugin/scenarios/httpclient-5.x-scenario/config/expectedData.yaml (hunk 1 above)
Comment (heihaozi) [reply test/plugin/scenarios/httpclient-5.x-scenario/config/expectedData.yaml:113]: OK, I add it.

File: apm-sniffer/apm-sdk-plugin/httpclient-5.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/httpclient/v5/HttpAsyncClientDoExecuteInterceptor.java (hunk 2 above)
Comment (heihaozi) [reply apm-sniffer/apm-sdk-plugin/httpclient-5.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/httpclient/v5/HttpAsyncClientDoExecuteInterceptor.java:43]: This is invalid code, I will remove it.

**Comment by wu-sheng** [comment]
> Update suggested change

You should not do this, I didn't change all related things. Please apply changes locally.

File: test/plugin/scenarios/httpclient-5.x-scenario/config/expectedData.yaml [hunk 4]
```diff
@@ -110,7 +110,7 @@ segmentItems:
             spanLayer: Http
             startTime: nq 0
             endTime: nq 0
-            componentId: not null
+            componentId: SpringMVC
```
Comment (wu-sheng) [inline test/plugin/scenarios/httpclient-5.x-scenario/config/expectedData.yaml:113]: This is ID, not name.

Comment (heihaozi) [reply test/plugin/scenarios/httpclient-5.x-scenario/config/expectedData.yaml:113]: OK, I fix it.

File: CHANGES.md [hunk 5]
```diff
@@ -26,6 +26,7 @@ Release Notes.
     rename `plugin.toolkit.log.grpc.reporter.max_message_size` to `log.max_message_size`.
 * Implement Kafka Log Reporter. Add config item of `agnt.conf`, `plugin.kafka.topic_logging`.
 * Upgrade byte-buddy to 1.11.18
+* Support for Apache HttpClient 5 plug.
```
Comment (wu-sheng) [inline CHANGES.md:29]:
```suggestion
* Add plugin to support Apache HttpClient 5.
```

**Review by wu-sheng — APPROVED** [review]

**Review by wankai123 — DISMISSED** [review]

File: test/plugin/scenarios/httpclient-5.x-scenario/config/expectedData.yaml [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    segments:
+      - segmentId: not null
+        spans:
+          - operationName: /httpclient-5.x/back
+            parentSpanId: -1
+            spanId: 0
+            spanLayer: Http
+            startTime: nq 0
+            endTime: nq 0
+            componentId: 14
```
Comment (wu-sheng) [inline test/plugin/scenarios/httpclient-5.x-scenario/config/expectedData.yaml:28]: As your controller in `org.apache.skywalking`, this wouldn't be traced by SpringMVC plugin, but Tomcat plugin should work.

```suggestion
            componentId: 1
```

File: test/plugin/scenarios/httpclient-5.x-scenario/config/expectedData.yaml [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+            skipAnalysis: 'false'
+      - segmentId: not null
+        spans:
+          - operationName: /httpclient-5.x/case/asyncGet
+            parentSpanId: -1
+            spanId: 0
+            spanLayer: Http
+            startTime: nq 0
+            endTime: nq 0
+            componentId: 14
```
Comment (wu-sheng) [inline test/plugin/scenarios/httpclient-5.x-scenario/config/expectedData.yaml:79]:
```suggestion
            componentId: 1
```

Same here.

File: test/plugin/scenarios/httpclient-5.x-scenario/config/expectedData.yaml (hunk 6 above)
Comment (wu-sheng) [reply test/plugin/scenarios/httpclient-5.x-scenario/config/expectedData.yaml:28]: I expect the tests will fail.

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**